metrics/README.md

```diff
@@ -0,0 +1,90 @@
+# Common metrics export interfaces for Knative
+
+_Note that this directory is currently in transition. See [the Plan](#the-plan)
+for details on where this is heading._
+
+## Current status
+
+The code currently uses OpenCensus to support exporting metrics to multiple
+backends. Currently, two backends are supported: Prometheus and Stackdriver.
+
+Metrics export is controlled by a ConfigMap called `config-observability` which
+is a key-value map with specific values supported for each of the Stackdriver
+and Prometheus backends. Hot-reload of the ConfigMap on a running process is
+supported by directly watching (via the Kubernetes API) the
+`config-observability` object. Configuration via environment is also supported
+for use by the `queue-proxy`, which runs with user permissions in the user's
+namespace.
+
+## Problems
+
+There are currently
+[6 supported Golang exporters for OpenCensus](https://opencensus.io/exporters/supported-exporters/go/).
+At least the Stackdriver exporter causes problems/failures if started without
+access to (Google) application defaut credentials. It's not clear that we want
+to build all of those backends into the core of `knative.dev/pkg` and all
+downstream dependents, and we'd like all the code shipped in `knative.dev/pkg`
+to be able to be tested without needing any special environment setup.
+
+With the current direct-integration setup, there needs to be initial and ongoing
+work in `pkg` (which should be high-value, low-churn code) to maintain and
+update stats exporters which need to be statically linked into ~all Knative
+binaries. This setup also causes problems for vendors who may want or need to
+perform an out-of-tree integration (e.g. proprietary or partially-proprietary
+monitoring stacks).
+
+Another problem is that each vendor's exporter requires different parameters,
+supplied as Golang `Options` methods which may require complex connections with
+the Knative ConfigMap. Two examples of this are secrets like API keys and the
+Prometheus monitoring port (which requires additional service/etc wiring).
+
+See also
+[this doc](https://docs.google.com/document/d/1t-aov3XrhobjCKW4kwScY44QAoahiwxoyXXFtZyL8jw/edit),
+where the plan was worked out.
+
+## The plan
+
+OpenCensus (and eventually OpenTelemetry) offers an sidecar or host-level agent
+with speaks the OpenCensus protocol and can proxy from this protocol to multiple
+backends.
+
+![OpenCensus Agent configuration](https://github.com/census-instrumentation/opencensus-service/raw/master/images/opencensus-service-deployment-models.png)
+
+**We will standardize on export to the OpenCensus agent, and encourage vendors
+to implement their own OpenCensus DaemonSet which connects to their desired
+monitoring environment.**
+
+**Additionally, once OpenTelemetry agent is stable, we will propose adding the
```

Comment:
Would Knative be willing to migrate to OpenTelemetry when they reach Beta, or would we wait for GA? I believe they are currently in Alpha with Beta around the corner.

On the collector service side, the OpenTelemetry collector service will actually support metrics instrumented using the OpenCensus library as well as the OpenTelemetry library. This gives us the nice benefit that we can switch from OpenCensus collector service -> OpenTelemetry collector service without modifying Knative code.

We'll need to migrate the Knative code to the OpenTelemetry library eventually, but we can consider the collector service migration and the library migration to be orthogonal issues.

Reply:
Rather than using an arbitrary cutoff like "Beta" or "GA" for OpenTelemetry for the agent, it seems like vendors could choose which they want to ship. As for the runtime contract, that seems like a nice addition, but it will be a "MAY" or eventually "SHOULD" but not a "MUST" for at least 6-9 months after introduction (and I'd only move to "MUST" if everyone participating agrees that they are already providing that service in a compatible way).

Reply:
For agents/collector services vendors can choose independently, but for the actual instrumentation libraries Knative will have to choose when to migrate from OpenCensus or OpenTelemetry.

Reply:
Are there different words I should put in here now?

I don't expect that we'll make a decision now about when Knative will switch to OpenTelemetry in the future -- I expect at some point we'll say "this seems stable enough, let's switch" as opposed to something date-driven.

Reply:
No, no need for wording change, just a general question :) 

Just trying to get a feel for when we would start investing in the migration work. I believe starting with OpenTelemetry Collector Service Beta sometime in the next couple months, OpenTelemetry will start recommending OpenTelemetry Collector Service over OpenCensus. That might be the turning point for starting to strip out vendor-specific code.

Reply:
I think once Google has a new export process that handles the Stackdriver things that are going on, we should start ripping out the non-`ocagent` backends.